backtracking/rat_in_maze.py

```diff
@@ -1,81 +1,125 @@
 from __future__ import annotations
 
 
-def solve_maze(maze: list[list[int]]) -> bool:
+def solve_maze(
+    maze: list[list[int]],
+    source_row: int,
+    source_column: int,
+    destination_row: int,
+    destination_column: int,
+) -> list[list[int]] | None:
     """
     This method solves the "rat in maze" problem.
-    In this problem we have some n by n matrix, a start point and an end point.
-    We want to go from the start to the end. In this matrix zeroes represent walls
-    and ones paths we can use.
     Parameters :
-        maze(2D matrix) : maze
+        - maze(2D matrix) : maze
+        - source_row (int): The row index of the starting point.
+        - source_column (int): The column index of the starting point.
+        - destination_row (int): The row index of the destination point.
+        - destination_column (int): The column index of the destination point.
     Returns:
-        Return: True if the maze has a solution or False if it does not.
+        Return: solution(2D matrix) if path exist ,otherwise None.
+    Description:
+        This method navigates through a maze represented as an n by n matrix,
+        starting from a specified source cell  and
+        aiming to reach a destination cell.
+        The maze consists of walls (1s) and open paths (0s).
+        By providing custom row and column values, the source and destination
+        cells can be adjusted.
     >>> maze = [[0, 1, 0, 1, 1],
     ...         [0, 0, 0, 0, 0],
     ...         [1, 0, 1, 0, 1],
     ...         [0, 0, 1, 0, 0],
     ...         [1, 0, 0, 1, 0]]
-    >>> solve_maze(maze)
-    [1, 0, 0, 0, 0]
-    [1, 1, 1, 1, 0]
-    [0, 0, 0, 1, 0]
-    [0, 0, 0, 1, 1]
-    [0, 0, 0, 0, 1]
-    True
+    >>> solve_maze(maze,0,0,len(maze)-1,len(maze)-1)
+    [[0, 1, 1, 1, 1], [0, 0, 0, 0, 1], [1, 1, 1, 0, 1],\
```

Comment:
Why is the output on the same line now?

Reply:
https://docs.python.org/3/library/doctest.html#doctest.NORMALIZE_WHITESPACE will allow you to have doctests that ignore the whitespace so the matrix can look like a matrix.